File: python-client/src/network/peer.py
import socket
import json
import hashlib
from pathlib import Path
import logging

logger = logging.getLogger(__name__)


class Peer:
    def __init__(self, peer_id, address):
        self.peer_id = peer_id
        self.address = address
        self.connected_peers = {}
        # Initialize shared directory - use ~/.p2p-share/shared as default
        self.shared_directory = Path.home() / '.p2p-share' / 'shared'
        self.shared_directory.mkdir(parents=True, exist_ok=True)
        logger.info(f"Using shared directory: {self.shared_directory}")

    def connect(self, peer):
        if peer.peer_id not in self.connected_peers:
            self.connected_peers[peer.peer_id] = peer
            logger.info(f"Connected to peer: {peer.peer_id} at {peer.address}")

    def disconnect(self, peer):
        if peer.peer_id in self.connected_peers:
            del self.connected_peers[peer.peer_id]
            logger.info(f"Disconnected from peer: {peer.peer_id}")

    def send_file_request(self, peer_id, file_name):
        if peer_id in self.connected_peers:
            logger.info(f"Requesting file '{file_name}' from peer: {peer_id}")
            try:
                peer = self.connected_peers[peer_id]
                response = peer.receive_file_request(self.peer_id, file_name)

                if response and isinstance(response, dict):
                    if response['status'] == 'approved':
                        # Create socket connection to peer
                        host, port = peer.address.split(':')
                        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                            s.connect((host, int(port)))

                            # Send file request
                            request = f"REQUEST_FILE:{file_name}"
                            s.sendall(request.encode())

                            # Receive file data
                            file_path = self.shared_directory / file_name

                            # First read response header
                            data = s.recv(1024)
                            if data.startswith(b"FILE_DATA:"):
                                # Parse header to get file details
                                header = data.decode(
                                    'utf-8', errors='ignore').split(':', 3)
                                if len(header) < 3:
                                    logger.error(
                                        "Invalid file header received")
                                    return False

                                # Extract filename and filesize
                                received_filename = header[1]
                                try:
                                    filesize = int(header[2])
                                except ValueError:
                                    logger.error(
                                        f"Invalid file size in header: {header[2]}")
                                    return False

                                logger.info(
                                    f"Receiving file: {received_filename} ({filesize} bytes)")

                                # Write any data already received after the header
                                header_size = len(
                                    "FILE_DATA") + len(received_filename) + len(str(filesize)) + 3
                                with open(file_path, 'wb') as f:
                                    if len(data) > header_size:
                                        f.write(data[header_size:])
                                        bytes_received = len(
                                            data) - header_size
                                    else:
                                        bytes_received = 0

                                    # Continue receiving the file
                                    while bytes_received < filesize:
                                        chunk = s.recv(4096)
                                        if not chunk:
                                            break
                                        f.write(chunk)
                                        bytes_received += len(chunk)

                                # Verify file hash if provided
                                if 'file_hash' in response:
                                    if self._verify_file_hash(file_path, response['file_hash']):
                                        logger.info(
                                            f"File '{file_name}' successfully received and verified")
                                        return True
                                    else:
                                        logger.error(
                                            "File verification failed")
                                        file_path.unlink()  # Delete corrupted file
                                        return False
                                else:
                                    logger.info(
                                        f"File '{file_name}' successfully received (no hash verification)")
                                    return True
                            elif data.startswith(b"ERR"):
                                error_msg = data.decode('utf-8').split(':', 1)
                                if len(error_msg) > 1:
                                    logger.error(
                                        f"Error from peer: {error_msg[1]}")
                                else:
                                    logger.error("Unknown error occurred")
                                return False
                            else:
                                logger.error(
                                    f"Unexpected response: {data[:20]}")
                                return False
                    else:
                        logger.info("File request denied by peer")
                        return False
                else:
                    logger.error("Invalid response from peer")
                    return False

            except Exception as e:
                logger.error(f"Error requesting file: {e}")
                return False
        else:
            logger.error(f"Peer {peer_id} not connected")
            return False

    def receive_file_request(self, peer_id, file_name):
        logger.info(
            f"Received file request for '{file_name}' from peer: {peer_id}")
        try:
            file_path = self.shared_directory / file_name
            # Also check if file exists in the current directory as a fallback
            current_dir_path = Path(file_name)

            if file_path.exists():
                path_to_use = file_path
            elif current_dir_path.exists():
                path_to_use = current_dir_path
            else:
                logger.info(f"File '{file_name}' not found")
                return {'status': 'not_found'}

            # Calculate file hash
            file_hash = self._calculate_file_hash(path_to_use)

            # Request user consent
            print(
                f"\nAllow {peer_id} to download {file_name}? (y/n): ", end="")
            consent = input()
            if consent.lower() == 'y':
                return {
                    'status': 'approved',
                    'file_hash': file_hash,
                    'file_size': path_to_use.stat().st_size,
                    'file_path': str(path_to_use)
                }
            else:
                logger.info(f"Request for file '{file_name}' denied by user")
                return {'status': 'denied'}

        except Exception as e:
            logger.error(f"Error handling file request: {e}")
            return {'status': 'error', 'message': str(e)}

    def list_connected_peers(self):
        return list(self.connected_peers.keys())

    def _calculate_file_hash(self, file_path):
        """Calculate SHA-256 hash of a file"""
        try:
            hash_obj = hashlib.sha256()
            with open(file_path, 'rb') as f:
                for chunk in iter(lambda: f.read(4096), b''):
                    hash_obj.update(chunk)
            return hash_obj.hexdigest()
        except Exception as e:
            logger.error(f"Error calculating file hash: {e}")
            return None

    def _verify_file_hash(self, file_path, expected_hash):
        """Verify that file hash matches expected hash"""
        actual_hash = self._calculate_file_hash(file_path)
        if not actual_hash:
            return False
        return actual_hash == expected_hash

File: go-client/internal/network/peer/peer.go
package peer

import (
	"crypto/rsa"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"os"
	"p2p-file-sharing/go-client/internal/crypto/auth"
	"p2p-file-sharing/go-client/internal/crypto/keys"
	"path/filepath"
	"sync"
)

// Peer represents a node in the P2P network
type Peer struct {
	PeerID     string
	Address    string
	PrivateKey *rsa.PrivateKey
	PublicKey  *rsa.PublicKey
	Auth       *auth.PeerAuthentication
	SharedDir  string

	// Connected peers
	mu             sync.RWMutex
	connectedPeers map[string]string // Map of peer IDs to addresses
}

// FileRequest represents a request for a file
type FileRequest struct {
	FileName  string `json:"file_name"`
	Requester string `json:"requester"`
}

// FileResponse represents a response to a file request
type FileResponse struct {
	Status   string `json:"status"`
	Reason   string `json:"reason,omitempty"`
	FileHash string `json:"file_hash,omitempty"`
	FileSize int64  `json:"file_size,omitempty"`
}

// NewPeer creates a new peer instance
func NewPeer(peerID, address, keysDir, sharedDir string) (*Peer, error) {
	// Load or generate keys
	privatePath := filepath.Join(keysDir, "private.pem")
	publicPath := filepath.Join(keysDir, "public.pem")

	var privateKey *rsa.PrivateKey
	var publicKey *rsa.PublicKey

	// Check if keys exist
	if _, err := os.Stat(privatePath); os.IsNotExist(err) {
		// Generate new keys
		keyPair, err := keys.GenerateKeyPair()
		if err != nil {
			return nil, fmt.Errorf("failed to generate key pair: %w", err)
		}

		privateKey = keyPair.PrivateKey
		publicKey = keyPair.PublicKey

		// Save keys
		if err := keys.SavePrivateKey(privateKey, privatePath); err != nil {
			return nil, fmt.Errorf("failed to save private key: %w", err)
		}

		if err := keys.SavePublicKey(publicKey, publicPath); err != nil {
			return nil, fmt.Errorf("failed to save public key: %w", err)
		}
	} else {
		// Load existing keys
		privateKey, err = keys.LoadPrivateKey(privatePath)
		if err != nil {
			return nil, fmt.Errorf("failed to load private key: %w", err)
		}

		publicKey, err = keys.LoadPublicKey(publicPath)
		if err != nil {
			return nil, fmt.Errorf("failed to load public key: %w", err)
		}
	}

	// Create authentication manager
	authManager := auth.NewPeerAuthentication(peerID, privateKey, publicKey)

	// Create shared directory if it doesn't exist
	if err := os.MkdirAll(sharedDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create shared directory: %w", err)
	}

	return &Peer{
		PeerID:         peerID,
		Address:        address,
		PrivateKey:     privateKey,
		PublicKey:      publicKey,
		Auth:           authManager,
		SharedDir:      sharedDir,
		connectedPeers: make(map[string]string),
	}, nil
}

// Connect establishes a secure connection with another peer
func (p *Peer) Connect(address string) error {
	// Parse address
	host, port, err := net.SplitHostPort(address)
	if err != nil {
		return fmt.Errorf("invalid address format: %w", err)
	}

	// Connect to peer
	conn, err := net.Dial("tcp", net.JoinHostPort(host, port))
	if err != nil {
		return fmt.Errorf("failed to connect to peer: %w", err)
	}
	defer conn.Close()

	// Initiate authentication
	authReq, err := p.Auth.InitiateAuthentication()
	if err != nil {
		return fmt.Errorf("failed to create auth request: %w", err)
	}

	// Serialize request
	authReqBytes, err := json.Marshal(authReq)
	if err != nil {
		return fmt.Errorf("failed to marshal auth request: %w", err)
	}

	// Send authentication request
	_, err = conn.Write(authReqBytes)
	if err != nil {
		return fmt.Errorf("failed to send auth request: %w", err)
	}

	// Read response
	buffer := make([]byte, 4096)
	n, err := conn.Read(buffer)
	if err != nil {
		return fmt.Errorf("failed to read auth response: %w", err)
	}

	// Parse response
	var authResp auth.AuthResponse
	if err := json.Unmarshal(buffer[:n], &authResp); err != nil {
		return fmt.Errorf("failed to parse auth response: %w", err)
	}

	// For now, we assume we know the peer's public key
	// In a real implementation, we would exchange public keys
	var peerPublicKey *rsa.PublicKey

	// Verify response
	if err := p.Auth.VerifyAuthResponse(&authResp, peerPublicKey); err != nil {
		return fmt.Errorf("auth response verification failed: %w", err)
	}

	// Add to verified peers
	p.Auth.AddVerifiedPeer(authResp.PeerID, peerPublicKey, authResp.DHPublicKey)

	// Add to connected peers
	p.mu.Lock()
	p.connectedPeers[authResp.PeerID] = address
	p.mu.Unlock()

	fmt.Printf("✅ Connected to peer %s at %s\n", authResp.PeerID, address)
	return nil
}

// RequestFile requests a file from a peer
func (p *Peer) RequestFile(peerID, fileName string) error {
	// Check if peer is connected
	p.mu.RLock()
	address, exists := p.connectedPeers[peerID]
	p.mu.RUnlock()

	if !exists {
		return errors.New("peer not connected")
	}

	// Create file request
	request := FileRequest{
		FileName:  fileName,
		Requester: p.PeerID,
	}

	// Serialize request
	requestBytes, err := json.Marshal(request)
	if err != nil {
		return fmt.Errorf("failed to marshal file request: %w", err)
	}

	// Encrypt request
	encryptedRequest, err := p.Auth.EncryptMessage(peerID, requestBytes)
	if err != nil {
		return fmt.Errorf("failed to encrypt file request: %w", err)
	}

	// Parse address
	host, port, err := net.SplitHostPort(address)
	if err != nil {
		return fmt.Errorf("invalid address format: %w", err)
	}

	// Connect to peer
	conn, err := net.Dial("tcp", net.JoinHostPort(host, port))
	if err != nil {
		return fmt.Errorf("failed to connect to peer: %w", err)
	}
	defer conn.Close()

	// Send request type
	_, err = conn.Write([]byte("FILE_REQUEST"))
	if err != nil {
		return fmt.Errorf("failed to send request type: %w", err)
	}

	// Read acknowledgment
	ack := make([]byte, 3)
	_, err = conn.Read(ack)
	if err != nil {
		return fmt.Errorf("failed to read acknowledgment: %w", err)
	}

	if string(ack) != "ACK" {
		return errors.New("invalid acknowledgment")
	}

	// Send encrypted request
	_, err = conn.Write(encryptedRequest)
	if err != nil {
		return fmt.Errorf("failed to send file request: %w", err)
	}

	// Read response
	buffer := make([]byte, 4096)
	n, err := conn.Read(buffer)
	if err != nil {
		return fmt.Errorf("failed to read response: %w", err)
	}

	// Decrypt response
	decryptedResponse, err := p.Auth.DecryptMessage(peerID, buffer[:n])
	if err != nil {
		return fmt.Errorf("failed to decrypt response: %w", err)
	}

	// Parse response
	var response FileResponse
	if err := json.Unmarshal(decryptedResponse, &response); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	// Check response status
	if response.Status != "approved" {
		reason := response.Reason
		if reason == "" {
			reason = "unknown"
		}
		return fmt.Errorf("file request denied: %s", reason)
	}

	// Receive file
	return p.receiveFile(conn, fileName, peerID, response.FileHash)
}

// receiveFile receives a file from a peer
func (p *Peer) receiveFile(conn net.Conn, fileName, peerID, expectedHash string) error {
	// Create file
	filePath := filepath.Join(p.SharedDir, fileName)
	file, err := os.Create(filePath)
	if err != nil {
		return fmt.Errorf("failed to create file: %w", err)
	}
	defer file.Close()

	// Send ready signal
	_, err = conn.Write([]byte("READY"))
	if err != nil {
		return fmt.Errorf("failed to send ready signal: %w", err)
	}

	// Receive file chunks
	hasher := sha256.New()

	for {
		// Read chunk size
		sizeBuffer := make([]byte, 8)
		_, err := io.ReadFull(conn, sizeBuffer)
		if err != nil {
			if err == io.EOF {
				break
			}
			return fmt.Errorf("failed to read chunk size: %w", err)
		}

		// Parse chunk size
		chunkSize := 0
		for i := 0; i < 8; i++ {
			chunkSize = (chunkSize << 8) | int(sizeBuffer[i])
		}

		// Check for end marker
		if chunkSize == 0 {
			break
		}

		// Read chunk
		chunk := make([]byte, chunkSize)
		_, err = io.ReadFull(conn, chunk)
		if err != nil {
			return fmt.Errorf("failed to read chunk: %w", err)
		}

		// Decrypt chunk (in a real implementation)
		// For simplicity, we'll skip actual decryption here

		// Update hash
		hasher.Write(chunk)

		// Write to file
		_, err = file.Write(chunk)
		if err != nil {
			return fmt.Errorf("failed to write to file: %w", err)
		}

		// Send acknowledgment
		_, err = conn.Write([]byte("OK"))
		if err != nil {
			return fmt.Errorf("failed to send acknowledgment: %w", err)
		}
	}

	// Verify file hash
	actualHash := fmt.Sprintf("%x", hasher.Sum(nil))
	if actualHash != expectedHash {
		// Remove file if hash doesn't match
		os.Remove(filePath)
		return fmt.Errorf("file hash mismatch: expected %s, got %s", expectedHash, actualHash)
	}

	fmt.Printf("✅ Successfully received file: %s\n", fileName)
	return nil
}

// HandleFileRequest processes an incoming file request
func (p *Peer) HandleFileRequest(conn net.Conn, encryptedRequest []byte, requesterID string) error {
	// Decrypt request
	decryptedRequest, err := p.Auth.DecryptMessage(requesterID, encryptedRequest)
	if err != nil {
		return fmt.Errorf("failed to decrypt request: %w", err)
	}

	// Parse request
	var request FileRequest
	if err := json.Unmarshal(decryptedRequest, &request); err != nil {
		return fmt.Errorf("failed to parse request: %w", err)
	}

	// Verify requester ID
	if request.Requester != requesterID {
		return errors.New("requester ID mismatch")
	}

	// Check if file exists
	filePath := filepath.Join(p.SharedDir, request.FileName)
	fileInfo, err := os.Stat(filePath)
	if err != nil {
		// Create and encrypt response
		response := FileResponse{
			Status: "denied",
			Reason: "file_not_found",
		}
		responseBytes, _ := json.Marshal(response)
		encryptedResponse, _ := p.Auth.EncryptMessage(requesterID, responseBytes)

		// Send response
		conn.Write(encryptedResponse)
		return errors.New("file not found")
	}

	// Calculate file hash
	fileHash, err := calculateFileHash(filePath)
	if err != nil {
		// Create and encrypt response
		response := FileResponse{
			Status: "denied",
			Reason: "hash_calculation_failed",
		}
		responseBytes, _ := json.Marshal(response)
		encryptedResponse, _ := p.Auth.EncryptMessage(requesterID, responseBytes)

		// Send response
		conn.Write(encryptedResponse)
		return fmt.Errorf("failed to calculate file hash: %w", err)
	}

	// Create response
	response := FileResponse{
		Status:   "approved",
		FileHash: fileHash,
		FileSize: fileInfo.Size(),
	}

	// Serialize and encrypt response
	responseBytes, err := json.Marshal(response)
	if err != nil {
		return fmt.Errorf("failed to marshal response: %w", err)
	}

	encryptedResponse, err := p.Auth.EncryptMessage(requesterID, responseBytes)
	if err != nil {
		return fmt.Errorf("failed to encrypt response: %w", err)
	}

	// Send response
	_, err = conn.Write(encryptedResponse)
	if err != nil {
		return fmt.Errorf("failed to send response: %w", err)
	}

	// Wait for ready signal
	readyBuffer := make([]byte, 5)
	_, err = conn.Read(readyBuffer)
	if err != nil {
		return fmt.Errorf("failed to read ready signal: %w", err)
	}

	if string(readyBuffer) != "READY" {
		return errors.New("invalid ready signal")
	}

	// Send file
	return p.sendFile(conn, filePath, requesterID)
}

// sendFile sends a file to a peer
func (p *Peer) sendFile(conn net.Conn, filePath, recipientID string) error {
	// Open file
	file, err := os.Open(filePath)
	if err != nil {
		return fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	// Send file in chunks
	buffer := make([]byte, 8192)

	for {
		// Read chunk
		n, err := file.Read(buffer)
		if err != nil {
			if err == io.EOF {
				break
			}
			return fmt.Errorf("failed to read file: %w", err)
		}

		// Encrypt chunk (in a real implementation)
		// For simplicity, we'll skip actual encryption here
		chunk := buffer[:n]

		// Send chunk size
		sizeBuffer := make([]byte, 8)
		for i := 0; i < 8; i++ {
			sizeBuffer[i] = byte((n >> (8 * (7 - i))) & 0xFF)
		}

		_, err = conn.Write(sizeBuffer)
		if err != nil {
			return fmt.Errorf("failed to send chunk size: %w", err)
		}

		// Send chunk
		_, err = conn.Write(chunk)
		if err != nil {
			return fmt.Errorf("failed to send chunk: %w", err)
		}

		// Wait for acknowledgment
		ackBuffer := make([]byte, 2)
		_, err = conn.Read(ackBuffer)
		if err != nil {
			return fmt.Errorf("failed to read acknowledgment: %w", err)
		}

		if string(ackBuffer) != "OK" {
			return errors.New("invalid acknowledgment")
		}
	}

	// Send end marker
	endMarker := make([]byte, 8)
	_, err = conn.Write(endMarker)
	if err != nil {
		return fmt.Errorf("failed to send end marker: %w", err)
	}

	fmt.Printf("✅ Successfully sent file: %s\n", filepath.Base(filePath))
	return nil
}

// MigrateKeys migrates to new keys for security
func (p *Peer) MigrateKeys() error {
	// Initiate key migration
	migrationData, err := p.Auth.InitiateKeyMigration()
	if err != nil {
		return fmt.Errorf("failed to initiate key migration: %w", err)
	}

	// Get new keys
	newPrivateKey := migrationData["new_private_key"].(*rsa.PrivateKey)
	newPublicKey := migrationData["new_public_key"].(*rsa.PublicKey)

	// For each connected peer, notify of key migration
	// This is simplified here - in a real implementation, we would
	// send a signed notification to each peer

	p.mu.RLock()
	fmt.Printf("🔄 Notifying %d connected peers of key migration...\n", len(p.connectedPeers))
	p.mu.RUnlock()

	// Activate new keys
	p.Auth.ActivateNewKeys(newPrivateKey, newPublicKey)
	p.PrivateKey = newPrivateKey
	p.PublicKey = newPublicKey

	// Save new keys
	homePath, _ := os.UserHomeDir()
	keysDir := filepath.Join(homePath, ".p2p-share-go", "keys")

	if err := keys.SavePrivateKey(newPrivateKey, filepath.Join(keysDir, "private.pem")); err != nil {
		return fmt.Errorf("failed to save new private key: %w", err)
	}

	if err := keys.SavePublicKey(newPublicKey, filepath.Join(keysDir, "public.pem")); err != nil {
		return fmt.Errorf("failed to save new public key: %w", err)
	}

	return nil
}

// ListConnectedPeers returns a list of connected peers
func (p *Peer) ListConnectedPeers() []string {
	p.mu.RLock()
	defer p.mu.RUnlock()

	peerIDs := make([]string, 0, len(p.connectedPeers))
	for peerID := range p.connectedPeers {
		peerIDs = append(peerIDs, peerID)
	}

	return peerIDs
}

// calculateFileHash computes a SHA-256 hash of a file
func calculateFileHash(filePath string) (string, error) {
	// Open file
	file, err := os.Open(filePath)
	if err != nil {
		return "", fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	// Calculate hash
	hasher := sha256.New()
	if _, err := io.Copy(hasher, file); err != nil {
		return "", fmt.Errorf("failed to read file: %w", err)
	}

	return fmt.Sprintf("%x", hasher.Sum(nil)), nil
}
